monai/application/datasets.py

```diff
@@ -0,0 +1,114 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
+import tarfile
+import numpy as np
+from typing import Any, Callable
+from monai.data import CacheDataset
+from monai.transforms import Randomizable
+from .utils import download_url, extractall
+
+
+class MedNISTDataset(Randomizable, CacheDataset):
+    """
+    The Dataset to automatically download MedNIST data and generate items for training, validation or test.
+    It's based on `CacheDataset` to accelerate the training process.
+
+    Args:
+        root: target dictionary to download and load MedNIST dataset.
+        section: expected data section, can be: `training`, `validation` or `test`.
+        transform: transforms to execute operations on input data.
+        download: whether to download the MedNIST from resource link, default is False.
+            if expected file already exists, skip downloading even set it to True.
+            user can manually copy `MedNIST.tar.gz` file or `MedNIST` folder to root directory.
+        extract: whether to extract the MedNIST.tar.gz file under root directory, default is False.
+        seed: random seed to randomly split training, validation and test datasets, defaut is 0.
+        cache_num: number of items to be cached. Default is `sys.maxsize`.
+            will take the minimum of (cache_num, data_length x cache_rate, data_length).
+        cache_rate: percentage of cached data in total, default is 1.0 (cache all).
+            will take the minimum of (cache_num, data_length x cache_rate, data_length).
+        num_workers: the number of worker threads to use.
+            If 0 a single thread will be used. Default is 0.
+
+    """
+
+    resource = "https://www.dropbox.com/s/5wwskxctvcxiuea/MedNIST.tar.gz?dl=1"
+    md5 = "0bc7306e7427e00ad1c5526a6677552d"
+
+    def __init__(
+        self,
+        root: str,
+        section: str,
+        transform: Callable[..., Any],
+        download: bool = False,
+        extract: bool = False,
+        seed: int = 0,
+        cache_num: int = sys.maxsize,
+        cache_rate: float = 1.0,
+        num_workers: int = 0,
+    ):
+        if not os.path.isdir(root):
+            raise ValueError("root must be a directory.")
+        self.section = section
+        self.set_random_state(seed=seed)
+        tarfile_name = os.path.join(root, "MedNIST.tar.gz")
+        dataset_dir = os.path.join(root, "MedNIST")
+        if download:
+            download_url(self.resource, tarfile_name, self.md5)
+        if extract:
```

Comment:
when download is False and extract is True, md5 should still be checked. This is related to line 54, why not use `download_and_extract` here?

Reply:
Do you mean to add a arg `download` to function `download_and_extract`?
Otherwise, if use below logic, it will skip `extracting` if download = False:
```py
if download:
    download_and_extract(... ...)
```

Reply:
Sorry I didn't make this clear. I think we always want to check the md5 before decompression, even if the tar file is fetched via some external methods. The current PR is not implemented in this way...

Reply:
OK, I added MD5 check before extracting in the latest commit.